EZP-30775: PHP API: Managing (#727)
ezsystems/developer-documentation: docs/api/public_php_api_managing_repository.md

# Managing the Repository

## Sections

[Sections](../guide/admin_panel.md#sections) enable you to divide content into groups
which can later be used e.g. as basis for permissions.

### Creating Sections

To create a new Section, you need to make use of the [`SectionCreateStruct`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/Values/Content/SectionCreateStruct.php)
and pass it to the [`SectionService::createSection`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/SectionService.php#L32) method:

``` php 
$sectionCreateStruct = $this->sectionService->newSectionCreateStruct();
$sectionCreateStruct->name = 'New section';
$sectionCreateStruct->identifier = 'newsection';
$this->sectionService->createSection($sectionCreateStruct);
```

### Getting Section information

You can use `SectionService` to retrieve Section information such as whether it is in use, or the total number of content assigned to it:

``` php

> $output->writeln(($this->sectionService->isSectionUsed($section) ? "This section is in use." : "This section is not in use."));

Reviewer: ```suggestion
$output->writeln(($this->sectionService->isSectionUsed($section) ? 'This section is in use.' : 'This section is not in use.'));
```

> $output->writeln("Number of Content items in this section: " . $this->sectionService->countAssignedContents($section));

Reviewer: ```suggestion
$output->writeln('Number of Content items in this section: ' . $this->sectionService->countAssignedContents($section));
```

```

### Listing content in a Section

To list Content items assigned to a Section you need to make a [query](php_public_api_search.md)

> for Content belonging to this section, by applying the [`SearchService`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/SearchService.php):

Reviewer 2: ```suggestion
for Content belonging to this section, by applying the [`SearchService`:](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/SearchService.php)
```

]);

$result = $this->searchService->findContentInfo($query);

foreach ($result->searchHits as $seachResult) {
    $output->writeln($seachResult->valueObject->name);
}
```

### Assigning Section to content

To assign content to a Section, use the [`SectionService::assignSection`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/SectionService.php#L111) method.
You need to provide it with the `ContentInfo` object of the Content item,
and the [`Section`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/Values/Content/Section.php) object:

``` php
$contentInfo = $this->contentService->loadContentInfo($contentId);
$section = $this->sectionService->loadSectionByIdentifier($sectionIdentifier);
$this->sectionService->assignSection($contentInfo, $section);
```

Note that assigning a Section to content does not automatically assign it to the Content item's children.

## Object states

[Object states](../guide/admin_panel.md#object-states)  enable you to set a custom state to any content.
States are grouped into Object state groups.

### Creating Object states

To create an Object state group and add Object states to it,

> you need to make use of the [`ObjectStateService`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/ObjectStateService.php):

Reviewer 2: ```suggestion
you need to make use of the [`ObjectStateService`:](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/ObjectStateService.php)
```

``` php
$objectStateGroupStruct = $this->objectStateService->newObjectStateGroupCreateStruct('rank');
$objectStateGroupStruct->defaultLanguageCode = 'eng-GB';
$objectStateGroupStruct->names = ['eng-GB' => 'rank'];
$this->objectStateService->createObjectStateGroup($objectStateGroupStruct);
```

[`ObjectStateService::createObjectStateGroup`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/ObjectStateService.php#L36)

> takes as argument an [`ObjectStateGroupCreateStruct`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/Values/ObjectState/ObjectStateGroupCreateStruct.php),

Reviewer 2: ```suggestion
takes as argument an [`ObjectStateGroupCreateStruct`,](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/Values/ObjectState/ObjectStateGroupCreateStruct.php)
```

$stateSpecialStruct->names = ['eng-GB' => 'special'];
$this->objectStateService->createObjectState($objectStateGroup, $stateSpecialStruct);
```

### Assigning Object state

To assign an Object state to a Content item,
use [`ObjectStateService::setContentState`.](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/ObjectStateService.php#L164)
Provide it with a `ContentInfo` object of the Content item, the Object state group and the Object state:

``` php
$contentInfo = $this->contentService->loadContentInfo($contentId);
$objectStateGroup = $this->objectStateService->loadObjectStateGroup($objectStateGroupId);
$objectState = $this->objectStateService->loadObjectState($objectStateId);

$this->objectStateService->setContentState($contentInfo, $objectStateGroup, $objectState);
```

!!! enterprise

    ## Workflow

    ### Getting workflow information

    To get information about a specific workflow for a Content item, use `WorkflowServiceInterface::loadWorkflowMetadataForContent`:

    ``` php
    $workflowMetadata = $this->workflowService->loadWorkflowMetadataForContent($content, $workflowName);
    foreach ($workflowMetadata->markings as $marking) {

>         $output->writeln($content->getName() . " is in stage " . $marking->name . " in workflow " . $workflowMetadata->workflow->getName());

Reviewer: ```suggestion
        $output->writeln($content->getName() . ' is in stage ' . $marking->name . ' in workflow ' . $workflowMetadata->workflow->getName());
```

    }
    ```

    !!! tip

        `marking`, a term from [Symfony Workflow,](https://symfony.com/doc/current/components/workflow.html)
        refers to a state in a workflow.

    To get a list of all workflows that can be used for a given Content item, use `WorkflowRegistry`:

    ``` php
    $supportedWorkflows = $this->workflowRegistry->getSupportedWorkflows($content);
    ```

    ### Applying workflow transitions

    To place a Content item in a workflow, use `WorkflowService::start`:

    ``` php
    $this->workflowService->start($content, $workflowName);
    ```

    To apply a transition to a Content item, use `WorkflowService::apply`.

>     Additionally, to you can check if the transition is possible for the given object using `WorkflowService::can`:

Reviewer 2: ```suggestion
    Additionally, you can check if the transition is possible for the given object using `WorkflowService::can`:
```

    ``` php
    if ($this->workflowService->can($workflowMetadata, $transitionName)) {
        $this->workflowService->apply($workflowMetadata, $transitionName, 'Please review');
    }
    ```

## Bookmarks

[`BookmarkService`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php)
enables you to read, add and remove bookmarks from content.

> To view a list of all bookmarks, use [`BookmarkService::loadBookmarks`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php#L54):

Reviewer 2: ```suggestion
To view a list of all bookmarks, use [`BookmarkService::loadBookmarks`:](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php#L54)
```

``` php
$bookmarkList = $this->bookmarkService->loadBookmarks();

> $output->writeln("Total bookmarks: " . $bookmarkList->totalCount);

Reviewer: ```suggestion
$output->writeln('Total bookmarks: ' . $bookmarkList->totalCount);
```

foreach ($bookmarkList->items as $bookmark) {
    $output->writeln($bookmark->getContentInfo()->name);
}
```

You can add a bookmark to a Content item by providing its Location object
to the [`BookmarkService::createBookmark`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php#L31) method:

``` php
$location = $this->locationService->loadLocation($locationId);

$this->bookmarkService->createBookmark($location);
```

> You can remove a bookmark from a Location with [`BookmarkService::deleteBookmark`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php#L42):

Reviewer 2: ```suggestion
You can remove a bookmark from a Location with [`BookmarkService::deleteBookmark`:](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/BookmarkService.php#L42)
```

``` php
$this->bookmarkService->deleteBookmark($location);
```

## Languages

### Getting Language information

> To get a list of all Languages in the system use [`LanguageService::loadLanguages`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/LanguageService.php#L81):

Reviewer 2: ```suggestion
To get a list of all Languages in the system use [`LanguageService::loadLanguages`:](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/LanguageService.php#L81)
```

``` php
$languageList = $this->languageService->loadLanguages();

foreach ($languageList as $language) {

>     $output->writeln($language->languageCode . ": " . $language->name);

Reviewer: ```suggestion
    $output->writeln($language->languageCode . ': ' . $language->name);
```

}
```

### Creating a Language

To create a new language, you need to create a [`LanguageCreateStruct`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/Values/Content/LanguageCreateStruct.php)
and provide it with the language code and language name.
Then, use [`LanguageService::createLanguage`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.3/eZ/Publish/API/Repository/LanguageService.php#L29) and pass the `LanguageCreateStruct` to it:

``` php
$languageCreateStruct = $this->languageService->newLanguageCreateStruct();

> $languageCreateStruct->languageCode = "ger-DE";

Reviewer: ```suggestion
$languageCreateStruct->languageCode = 'ger-DE';
```

> $languageCreateStruct->name = "German";

Reviewer: ```suggestion
$languageCreateStruct->name = 'German';
```
